NMS-14281: Add instructions for managing configuration files with git (#4962)
OpenNMS/opennms: docs/modules/deployment/pages/upgrade/git-upgrade.adoc

[[git-upgrade]]
= Upgrade {page-component-title} with Git

> This tutorial was originally written on the OpenNMS Discourse forum (see https://opennms.discourse.group/t/managing-configuration-changes-with-git/2327[Managing Configuration Changes with Git]).

Reviewer: ```suggestion
This content was originally written on the OpenNMS Discourse forum (see https://opennms.discourse.group/t/managing-configuration-changes-with-git/2327[Managing Configuration Changes with Git]).
Follow the directions here to upgrade with Git if you xref:deployment/upgrade/git-diff.adoc/ #git-diff[use Git to manage configuration changes].

```

IMPORTANT: Remember to stop {page-component-title} before beginning an upgrade.

== Commit configuration changes

Run the `git status` command to check if changes to the `etc` directory have been committed.

> If files have been modified, follow the instructions in xref:deployment:upgrade/git-diff.adoc#make-changes[Make your changes] to commit them.

Reviewer: For my own understanding ... is this a backup of the directory content? We revert to a pristine state, so I want to make sure they aren't losing anything.

> If files have been modified, follow the instructions in xref:deployment:upgrade/git-diff.adoc#make-changes[Make your changes] to commit them.

Author: From what I understand, that's correct. It's a final check for any modified files before an upgrade; if any files have been changed since the user's last commit, they should refer to the "Make changes" section to commit them. :) 

If there's a clearer way to word it, please let me know!

== Upgrade {page-component-title}

Upgrade your {page-component-title} installation as usual (see xref:deployment:upgrade/basic.adoc[Basic Upgrade Steps]).

After upgrading, run the `git status` command to see any changes to the default configuration files from your repository's `master` branch.
If any files were removed during the upgrade, run the `git rm <filename>` command.
This removes the files from your repository.

Make sure your repository's `master` branch is aligned with the version of {page-component-title} that you upgraded to by adding and committing the changes:

[source, console]
----
git add .
git commit -m 'Upgraded to OpenNMS x.x.x.'
----

Replace the `x.x.x` in the last line with your current {page-component-title} version.

When your `master` branch has been brought up to date, you can switch back to your modified branch:

[source, console]
----
git checkout local-modifications
----

== Apply changes from `master` branch

Update your modified branch by merging your `master` branch into your current working branch (in this example, `local-modifications`):

[source, console]
----
git merge master
----

> If git finds a conflict, the `merge` command will generate an error message:

Reviewer: ```suggestion
If Git finds a conflict, the `merge` command generates an error message:
```

[[git-upgrade]]
= Upgrade {page-component-title} with Git

This content was originally written on the OpenNMS Discourse forum (see https://opennms.discourse.group/t/managing-configuration-changes-with-git/2327[Managing Configuration Changes with Git]).
Follow the directions here to upgrade with Git if you xref:deployment:upgrade/git-diff.adoc#git-diff[use Git to manage configuration changes].
The steps below assume you have already set up your xref:deployment:upgrade/git-diff.adoc[configuration folders as a Git repository].

IMPORTANT: Remember to stop {page-component-title} before beginning an upgrade.

== Commit configuration changes

Run the `git status` command to check if changes to the `etc` directory have been committed.
If files have been modified, follow the instructions in xref:deployment:upgrade/git-diff.adoc#make-changes[Make changes] to commit them.

== Switch to pristine branch

Run this code to switch to your repository's `master` branch:

[source, console]
----

> git checkout master

Author: ```suggestion
git checkout main
```

----

This reverts your `etc` directory to its pristine condition in preparation for the upgrade.

== Upgrade {page-component-title}

Upgrade your {page-component-title} installation as usual (see xref:deployment:upgrade/basic.adoc[Basic Upgrade Steps]).

> After upgrading, run the `git status` command to see any changes to the default configuration files from your repository's `master` branch.

Author: ```suggestion
After upgrading, run the `git status` command to see any changes to the default configuration files from your repository's `main` branch.
```

If any files were removed during the upgrade, run the `git rm <filename>` command.
This removes the files from your repository.

> Make sure your repository's `master` branch is aligned with the version of {page-component-title} that you upgraded to by adding and committing the changes:

Author: ```suggestion
Make sure your repository's `main` branch is aligned with the version of {page-component-title} that you upgraded to by adding and committing the changes:
```

[source, console]
----
git add .
git commit -m 'Upgraded to OpenNMS x.x.x.'
----

Replace the `x.x.x` in the last line with your current {page-component-title} version.

> When your `master` branch has been brought up to date, you can switch back to your modified branch:

Author: ```suggestion
When your `main` branch has been brought up to date, you can switch back to your modified branch:
```

[source, console]
----
git checkout local-modifications
----

> == Apply changes from `master` branch

Author: ```suggestion
== Apply changes from `main` branch
```

> Update your modified branch by merging your `master` branch into your current working branch (in this example, `local-modifications`):

Author: ```suggestion
Update your modified branch by merging your `main` branch into your current working branch (in this example, `local-modifications`):
```

[source, console]
----

> git merge master

Author: ```suggestion
git merge main
```

----

If Git finds a conflict, the `merge` command generates an error message:

[source, console]
----

> git merge master

Author: ```suggestion
git merge main
```
